gamd/gamdSimulation.py

```diff
@@ -112,25 +112,44 @@ def createGamdSimulation(self, config, platform_name, device_index):
                 raise Exception("Invalid input type: %s. Allowed types are: "\
                                 "'pdb' and 'rst7'/'inpcrd'.")
             gamdSimulation.system = prmtop.createSystem(
-                nonbondedMethod=nonbondedMethod, 
-                nonbondedCutoff=config.system.nonbonded_cutoff, 
+                nonbondedMethod=nonbondedMethod,
+                nonbondedCutoff=config.system.nonbonded_cutoff,
                 constraints=constraints)
-            
+
+
         elif config.input_files.charmm is not None:
-            psf = openmm_app.CharmmPsfFile(
-                config.input_files.charmm.topology)
-            pdb_coords_filename = config.input_files.charmm.coordinates
-            positions, box_vectors = load_pdb_positions_and_box_vectors(
-                pdb_coords_filename, need_box)
+            psf = openmm_app.CharmmPsfFile(config.input_files.charmm.topology)
+            if config.input_files.charmm.coordinates_filetype == "crd":
+                positions = openmm_app.CharmmCrdFile(
+                                          config.input_files.charmm.coordinates)
+            elif config.input_files.charmm.coordinates_filetype == "pdb":
+                positions = openmm_app.PDBFile(
+                                          config.input_files.charmm.coordinates)
+            else:
+                raise Exception("Invalid input type: %s. Allowed types are: "\
+                                "'crd' and 'pdb'.")
+
+            boxinfo = config.input_files.charmm.box_vectors
+            boxlx = boxinfo["a"]
+            boxly = boxinfo["b"]
+            boxlz = boxinfo["c"]
+            alpha = boxinfo["alpha"]
+            beta  = boxinfo["beta"]
+            gamma = boxinfo["gamma"]
+            psf.setBox(boxlx, boxly, boxlz, alpha, beta, gamma)
```

Comment:
I believe this code makes an assumption that tags for a, b, and c exist.  Based on what I read in the parser, it is quite possible for them to not exist and we are just using a default.  I assume that means that setBox wouldn't be called at all?

I was thinking that the best way to handle this might be to add the following pieces of code:

in config.py in the CharmmConfig, add the following methods and variables:

self.config_box_vector_defined = False

def get_box_vectors(self):
    errorMessage = "The box vectors were only partially defined.  " \
                   "At least a, b, and c must be present."
    if not self.is_whole_box_defined():
        raise RuntimeError(errorMessage)
    return [self.box_vectors["a"], self.box_vectors["b"],
            self.box_vectors["c"],
            self.alpha, self.beta, self.gamma]


def is_whole_box_defined(self):
    lengths = ["a", "b", "c"]
    result = True
    for length in lengths:
        result = result and length in self.box_vectors
    return result


def is_config_box_vector_defined(self):
  return self.config_box_vector_defined

in parser.py, add the following once we are in the box-vector tag:

self.config_box_vector_defined = True

Then, right here, we should be able to just change it to something like the following:

if config.input_files.charmm.is_config_box_vector_defined:
    psf.setBox(*config.input_files.charmm.get_box_vectors())

This way, if they user doesn't define a box vector, we aren't making the call.  Of course, if the box vector is required for psf files, then we should throw an error instead if they didn't define it.  

FYI, I have no idea if I got that code exactly correct for the calling convention, but hopefully it gets the idea across.  What do you think?

Reply:
@dyelar the box lengths are required to create the system object with psf files, so they must be defined by the user. Currently, we throw an error if any of the lengths "a", "b" or "c" are not provided (parser.py, line 222). We do, however, make an assumption that the box angles are all 90 degrees if users do not provide alpha, beta or gamma. I can either raise an error or I can raise a warning on that matter. What do you prefer?